python/lsst/afw/geom/_hpxUtils.py

```diff
@@ -0,0 +1,400 @@
+
+    md = PropertySet()
+    md['CD1_1'] = -cos45_scale
+    md['CD1_2'] = -cos45_scale
+    md['CD2_1'] = cos45_scale
+    md['CD2_2'] = -cos45_scale
+    md['CTYPE1'] = 'RA---HPX'
+    md['CTYPE2'] = 'DEC--HPX'
+    md['CRVAL1'] = 0.0
+    md['CRVAL2'] = 0.0
+    md['PV2_1'] = 4
+    md['PV2_2'] = 3
+    md['CRPIX1'] = ((tilepix + 1)/2.0) - 0.5*(-cent_ra/cos45_scale + cent_dec/cos45_scale)
+    md['CRPIX2'] = ((tilepix + 1)/2.0) - 0.5*(-cent_ra/cos45_scale - cent_dec/cos45_scale)
+
+    return makeSkyWcs(md)
+
+
+def _hpx_center(order, pixel, tilepix):
+    """
+    Compute the projected healpix center for use in HPX headers.
+
+    Code is adapted from AladinSrc.jar version 11.024, Tile2HPX.java.
+
+    Parameters
+    ----------
+    order : `int`
+        Healpix order, such that nside=2**order.
+    pixel : `int`
+        Pixel number in the nest ordering scheme.
+    tilepix : `int`
+        Number of sub-pixels on the side of the pixel.
+        Must be a power of 2.
+
+    Returns
+    -------
+    cent_ra, cent_dec : `float`
+        Projected center ra/dec in degrees.
+
+    Raises
+    ------
+    `ValueError`: Raised if tilepix is not a power of 2,
+        or if order is <=0, or if pixel number is out of
+        range for the given order (0 < 12*nside*nside).
+    """
+    if (tilepix & (tilepix - 1) != 0) or tilepix <= 0:
+        raise ValueError(f"tilepix {tilepix} must be a power of 2.")
+
+    if order <= 0:
+        raise ValueError(f"order {order} must be positive.")
+    nside_tile = 2**order
+
+    if pixel < 0 or pixel >= 12*nside_tile*nside_tile:
+        raise ValueError(f"pixel value {pixel} out of range.")
+
+    twice_depth = np.left_shift(order, 1)
+    xy_mask = np.left_shift(1, twice_depth) - 1
+    fc = _ZOrderCurve2DInt()
+
+    d0h = np.int32(np.right_shift(pixel, twice_depth))
+    _hash = fc.hash2ij(pixel & xy_mask)
+    i_in_d0h = fc.ij2i(_hash)
+    j_in_d0h = fc.ij2j(_hash)
+    # Compute coordinates from the center of the base pixel
+    # with x-axis = W-->E, y-axis = S-->N
+    l_in_d0h = i_in_d0h - j_in_d0h
+    h_in_d0h = i_in_d0h + j_in_d0h - (nside_tile - 1)
+    # Compute coordinates of the base pixel in the projection plane
+    d0h_by_4_quotient = np.right_shift(d0h, 2)
+    d0h_mod_4 = d0h - np.left_shift(d0h_by_4_quotient, 2)
+    h_d0h = 1 - d0h_by_4_quotient
+    l_d0h = np.left_shift(d0h_mod_4, 1)
+    if ((h_d0h == 0) and ((l_d0h == 6) or ((l_d0h == 4) and (l_in_d0h > 0)))):
+        # Equatorial region
+        l_d0h -= 8
+    elif (h_d0h != 0):
+        # Polar caps regions
+        l_d0h += 1
+        if (l_d0h > 3):
+            l_d0h -= 8
+    # Finalize
+    return (np.rad2deg((np.pi/4.)*(l_d0h + l_in_d0h/float(nside_tile))),
+            np.rad2deg((np.pi/4.)*(h_d0h + h_in_d0h/float(nside_tile))))
+
+
+class _ZOrderCurve2DInt(object):
+    """
+    Z-Order 2D curve for 32-bit integer values.
+
+    Code is ported from AladinSrc.jar version 11.024,
```

Comment:
What is the code license for the aladin code?

Reply:
GPL3, which is compatible with us, right?  Where do I put that?

Reply:
Yes. We might want to note that explicitly in the comment because it effectively blocks us from changing the afw license in the future. Maybe even an explicit note at the top of the file so we won't miss it.

Reply:
I have added links to http://aladin.u-strasbg.fr/COPYING .  But the file is already GPLv3 as noted at the top, so I think that there's nothing else to add?

Reply:
That's great. The point is that we continue to ponder license changes to BSD 3 clause in parts of the code and this would put an explicit brake on that for afw.

Reply:
FYI, if that's possibly a problem for the future, I can put this routine elsewhere.